compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -388,28 +399,46 @@ impl StructType {
                     })
                     .collect();
 
-                (typ.substitute(&substitutions), i)
+                (field.typ.substitute(&substitutions), field.visibility, i)
             },
         )
     }
 
     /// Returns all the fields of this type, after being applied to the given generic arguments.
+    pub fn get_fields_with_visibility(
+        &self,
+        generic_args: &[Type],
+    ) -> Vec<(String, ItemVisibility, Type)> {
+        let substitutions = self.get_fields_subtitutions(generic_args);
```

Comment:
```suggestion
        let substitutions = self.get_fields_substitutions(generic_args);
```
And a couple other places